RavenDB-17760 : Sharding - Storage - Create a new tree to track bucket statistics

aviv86 opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-17760/Sharding-Storage-Prepare-storage-to-support-sharding

### Additional description

- add new `BucketStats` tree which will hold buckets info : size + number of docs in bucket + last modified date
- extend `DynamicKeyIndexDef` to allow to do additional work after insert/delete/update of indexed item
- use our "BucketAndEtag" dynamic-key indexes to update bucket info
- track bucket info in a in-memory dictionary and write it to disk just before commit
- provide a way to get stats for a specific bucket and to iterate over bucket stats + basic tests
- update the bucket stats directly (not via dynamic-index) on put/delete of attachment stream

### Type of change

- New feature

### How risky is the change?

- Moderate 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works
- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2571,6 +2576,52 @@ internal static ByteStringContext.Scope GenerateBucketAndEtagIndexKeyForTombston
             return GenerateBucketAndEtagIndexKey(context, idIndex: (int)TombstoneTable.LowerId, etagIndex: (int)TombstoneTable.Etag, ref tvr, out slice);
         }
 
+        // TODO : use a dedicated attribute
```
what do you mean, attribute?

aviv86 replied:
removed

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2571,6 +2576,52 @@ internal static ByteStringContext.Scope GenerateBucketAndEtagIndexKeyForTombston
+
+            bucketStats.Size += newSize - oldSize;
+
+            if (isDocument)
```
let's pass a `int numOfDocsChanged` and increment / decrement on that, instead of the `isDocument`

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2571,6 +2576,52 @@ internal static ByteStringContext.Scope GenerateBucketAndEtagIndexKeyForTombston
+
+        private static void UpdateBucketStats(Transaction tx, Slice key, int oldSize, int newSize, bool isDocument)
+        {
+            var nowTicks = DateTime.UtcNow.Ticks;
```
We should use the db time, instead

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2571,6 +2576,52 @@ internal static ByteStringContext.Scope GenerateBucketAndEtagIndexKeyForTombston
+                    var readResult = bucketStatsTree.Read(keySlice);
+                    if (readResult != null)
+                        bucketStats = *(Voron.Data.BucketStats*)readResult.Reader.Base;
+                }
```
Why do you need to do that?
You are doing that to read the number of docs / size that is already stored, no?
But why bother do that here, just aggregate the results and do the sum on commit.

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs:
```diff
@@ -39,6 +38,69 @@ public IEnumerable<Document> GetDocumentsByBucketFrom(DocumentsOperationContext
+    private static BucketStatistics ValueReaderToBucketStats(Slice key, ref ValueReader valueReader)
+    {
+        var bucket = Bits.SwapBytes(*(int*)key.Content.Ptr);
+        var stats = *(Voron.Data.BucketStats*)valueReader.Base;
```
Why `BucketStatistics ` vs `BucketStats` ?

ayende replied:
and why do we even have `BucketStats` inside Voron?

aviv86 replied:
moved `BucketStats` out of Voron

karmeli87 commented on src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs:
```diff
@@ -39,6 +38,69 @@ public IEnumerable<Document> GetDocumentsByBucketFrom(DocumentsOperationContext
+            Bucket = bucket,
+            Size = stats.Size,
+            NumberOfItems = stats.NumberOfDocuments,
+            LastModified = new DateTime(stats.LastModifiedTicks)
```
let's specify here and on all other places the `Utc` kind, just to be on the safe side

aviv86 replied:
done

karmeli87 commented on src/Voron/Data/BucketStats.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    [StructLayout(LayoutKind.Explicit)]
+    internal struct BucketStats
+    {
```
Let's put few reserves here? at least round this to 32 bits

karmeli87 commented on src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs:
```diff
@@ -348,6 +363,8 @@ public void Validate()
                 if (GenerateKey.Method.GetCustomAttribute<StorageIndexEntryKeyGeneratorAttribute>() == null)
                     throw new ArgumentOutOfRangeException(nameof(GenerateKey), $"{nameof(GenerateKey)} must be marked with custom attribute '{nameof(StorageIndexEntryKeyGeneratorAttribute)}'");
+
+                //  TODO : validate 'OnEntryChanged'
```
Keep in mind that I already plan to use it for my own things.. like to keep track of the last change vector.. 
Let's assume I will add this functionally after the release, what needs to be done, to ensure proper schema upgrade?

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2571,6 +2576,52 @@ internal static ByteStringContext.Scope GenerateBucketAndEtagIndexKeyForTombston
+            UpdateBucketStats(tx, key, oldSize, newSize, isDocument: false);
+        }
+
+        private static void UpdateBucketStats(Transaction tx, Slice key, int oldSize, int newSize, bool isDocument)
```
What happens to the size with documents compression?

aviv86 replied:
not working as expected, we count the Raw size and not the Compressed size
added a test (skipped for now) and created an issue : https://issues.hibernatingrhinos.com/issue/RavenDB-19530/

ayende commented on src/Voron/Impl/Transaction.cs:
```diff
@@ -281,6 +279,22 @@ internal void PrepareForCommit()
                 }
             }
 
+            if (_bucketStatistics != null)
+            {
+                var tree = ReadTree("BucketStats"); // TODO 
+                foreach ((int bucket, BucketStats stats) in _bucketStatistics)
+                {
+                    using (Allocator.Allocate(sizeof(int), out var keyBuffer))
+                    {
+                        *(int*)keyBuffer.Ptr = bucket;
+                        var keySlice = new Slice(keyBuffer);
+
+                        using (tree.DirectAdd(keySlice, sizeof(BucketStats), out byte* ptr))
+                            *(BucketStats*)ptr = stats;
+                    }
+                }
+            }
```
Make this an event outside of Voron

aviv86 replied:
done

aviv86 committed:
RavenDB-17760 : minor changes after review

aviv86 committed:
RavenDB-17760 : handle Serialization/ReadFrom/Validation of DynamicKeyIndexDef.OnEntryChanged

aviv86 committed:
RavenDB-17760 : add tests for DynamicKeyIndexDef.OnEntryChanged

aviv86 committed:
RavenDB-17760 : bucket stats - take into account attachment stream size - update bucket stats directly when adding a new stream or deleting an existing stream

aviv86 committed:
RavenDB-17760 : test bucket stats with documents compression enabled (not working as expected - skipped the test for now)

aviv86 committed:
RavenDB-17760 : fix tests failing on 'AmbiguousMatchException' in 'DynamicKeyIndexDef.ReadFrom' - have just a single method named 'UpdateBucketStats' in ShardedDocumentsStorage

aviv86 committed:
RavenDB-17760 - use swapped bucket as indexing key in BucketStats tree and provide a way to iterate over bucket stats entries

aviv86 committed:
RavenDB-17760 : basic tests

aviv86 committed:
RavenDB-17760 : address review comments :
- move BucketStats struct out of Voron
- keep the in-memory bucket-stats inside document storage instead of Voron
- add new 'Owner' object to Transaction.cs
- set Transaction.Owner to DocumentDatabase instance if we're in a DocumentsTransaction, so that we can accesses the in-memory stats structure from static DocumentsStorage.UpdateBucketStats
- add new OnBeforeCommit events to StorageEnvironment and LowLevelTransaction, and use it to update the BucketStatsTree before commit
- when updating stats - use db time instead of UtcNow
- change GetBucketStatistics params from 'start' and 'take' to 'fromBucket' and 'toBucket'

aviv86 committed:
RavenDB-17760 : minor modifications :
- move event OnBeforeCommit from StorageEnvironment to DocumentStorage,
- set OnBeforeCommit to UpdateBucketStatsTreeBeforeCommit from ShardedDocumentsStorage ctor
- append OnBeforeCommit to LowLevelTransaction.OnBeforeCommit from DocumentsTransaction ctor
- move all bucket stats handling from DocumentStorage to ShardedDocumentsStorage

aviv86 committed:
RavenDB-17760 : move tests to septate class, test bucket-stats with document extensions, cleanup

aviv86 committed:
RavenDB-17760 : minor fixes after rebase

aviv86 committed:
RavenDB-17760 : test bucket stats after bucket migration

karmeli87 commented on src/Raven.Server/Documents/AttachmentsStorage.cs:
```diff
@@ -1359,8 +1363,37 @@ public static IEnumerable<BlittableJsonReaderObject> GetAttachmentsFromDocumentM
                 (int)AttachmentsTable.Etag, ref tvr, out slice);
         }
+
+        private void UpdateBucketStatsOnPutOrDeleteStream(DocumentsOperationContext context, Slice attachmentKey, long sizeChange)
+        {
+            if (_documentDatabase is not ShardedDocumentDatabase)
+                return;
+
+            using (GetBucketFromAttachmentKey(context, attachmentKey, out var bucketSlice))
+            {
+                ShardedDocumentsStorage.UpdateBucketStats(context.Transaction.InnerTransaction, bucketSlice, oldSize : 0, newSize : sizeChange);
+            }
+        }
+
+        private static ByteStringContext.Scope GetBucketFromAttachmentKey(DocumentsOperationContext context, Slice attachmentKey, out Slice bucketSlice)
+        {
+            int sizeOfDocId = 0;
+            for (; sizeOfDocId < attachmentKey.Size; sizeOfDocId++)
+            {
+                if (attachmentKey[sizeOfDocId] == SpecialChars.RecordSeparator)
+                    break;
+            }
+
+            var bucket = ShardHelper.GetBucket(attachmentKey.Content.Ptr, sizeOfDocId);
+            var scope = context.Allocator.Allocate(sizeof(int), out var buffer);
+            *(int*)buffer.Ptr = Bits.SwapBytes(bucket);
+
+            bucketSlice = new Slice(buffer);
+
+            return scope;
+        }
```
let's move this and all sharded related functions into a new partial class `AttachmentsStorage.Sharding`

aviv86 replied:
done

karmeli87 commented on src/Raven.Server/Documents/ConflictsStorage.cs:
```diff
@@ -841,7 +841,7 @@ public string GetCollection(DocumentsOperationContext context, string id)
         internal static ByteStringContext.Scope GenerateBucketAndEtagIndexKeyForConflicts(ByteStringContext context, ref TableValueReader tvr, out Slice slice)
         {
-            return GenerateBucketAndEtagIndexKey(context, idIndex: (int)ConflictsTable.LowerId, etagIndex: (int)ConflictsTable.Etag, ref tvr, out slice);
+            return ShardedDocumentsStorage.GenerateBucketAndEtagIndexKey(context, idIndex: (int)ConflictsTable.LowerId, etagIndex: (int)ConflictsTable.Etag, ref tvr, out slice);
```
let's move this and all sharded related functions into a new partial class `ConflictsStorage.Sharding`

aviv86 replied:
done

karmeli87 commented on src/Raven.Server/Documents/ConflictsStorage.cs:
```diff
@@ -841,7 +841,7 @@ public string GetCollection(DocumentsOperationContext context, string id)
         internal static ByteStringContext.Scope GenerateBucketAndEtagIndexKeyForConflicts(ByteStringContext context, ref TableValueReader tvr, out Slice slice)
         {
-            return GenerateBucketAndEtagIndexKey(context, idIndex: (int)ConflictsTable.LowerId, etagIndex: (int)ConflictsTable.Etag, ref tvr, out slice);
+            return ShardedDocumentsStorage.GenerateBucketAndEtagIndexKey(context, idIndex: (int)ConflictsTable.LowerId, etagIndex: (int)ConflictsTable.Etag, ref tvr, out slice);
```
let's move this and all sharded related functions into a new partial class `ConflictsStorage.Sharding`

aviv86 replied:
done

karmeli87 commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -2429,7 +2429,7 @@ public long GetLastCounterEtag(DocumentsOperationContext context, string collect
         internal static ByteStringContext.Scope GenerateBucketAndEtagIndexKeyForCounters(ByteStringContext context, ref TableValueReader tvr, out Slice slice)
         {
-            return ExtractIdFromKeyAndGenerateBucketAndEtagIndexKey(context, (int)CountersTable.CounterKey, (int)CountersTable.Etag, ref tvr, out slice);
+            return ShardedDocumentsStorage.ExtractIdFromKeyAndGenerateBucketAndEtagIndexKey(context, (int)CountersTable.CounterKey, (int)CountersTable.Etag, ref tvr, out slice);
```
same

aviv86 replied:
done

karmeli87 commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -57,14 +57,18 @@ public unsafe class DocumentsStorage : IDisposable
+        public Action<LowLevelTransaction> OnBeforeCommit { get; protected set; }
+
+        protected Dictionary<string, CollectionName> _collectionsCache;
+        protected static readonly Slice BucketStatsSlice;
```
move to `ShardedDocumentsStorage`

aviv86 replied:
done

aviv86 replied:
done

karmeli87 commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -87,6 +91,7 @@ static DocumentsStorage()
                 Slice.From(ctx, "GlobalTree", ByteStringType.Immutable, out GlobalTreeSlice);
                 Slice.From(ctx, "GlobalChangeVector", ByteStringType.Immutable, out GlobalChangeVectorSlice);
                 Slice.From(ctx, "GlobalFullChangeVector", ByteStringType.Immutable, out GlobalFullChangeVectorSlice);
+                Slice.From(ctx, "BucketStats", ByteStringType.Immutable, out BucketStatsSlice);
```
move to ShardedDocumentsStorage

aviv86 replied:
done

karmeli87 commented on src/Raven.Server/Documents/Revisions/RevisionsStorage.cs:
```diff
@@ -1960,7 +1960,7 @@ public long GetNumberOfRevisionDocuments(DocumentsOperationContext context)
         internal static ByteStringContext.Scope GenerateBucketAndEtagIndexKeyForRevisions(ByteStringContext context, ref TableValueReader tvr, out Slice slice)
         {
-            return GenerateBucketAndEtagIndexKey(context, idIndex: (int)RevisionsTable.LowerId, etagIndex: (int)RevisionsTable.Etag, ref tvr, out slice);
+            return ShardedDocumentsStorage.GenerateBucketAndEtagIndexKey(context, idIndex: (int)RevisionsTable.LowerId, etagIndex: (int)RevisionsTable.Etag, ref tvr, out slice);
```
RevisionsStorage.Sharding

aviv86 replied:
done

karmeli87 commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs:
```diff
@@ -2753,14 +2753,14 @@ private static void EnsureStatsAndDataIntegrity(DocumentsOperationContext contex
         internal static ByteStringContext.Scope GenerateBucketAndEtagIndexKeyForTimeSeries(ByteStringContext context, ref TableValueReader tvr, out Slice slice)
         {
-            return DocumentsStorage.ExtractIdFromKeyAndGenerateBucketAndEtagIndexKey(context, keyIndex: (int)TimeSeriesTable.TimeSeriesKey,
+            return ShardedDocumentsStorage.ExtractIdFromKeyAndGenerateBucketAndEtagIndexKey(context, keyIndex: (int)TimeSeriesTable.TimeSeriesKey,
```
same

aviv86 replied:
done

aviv86 committed:
RavenDB-17760 : address PR comments - move all sharded related functions into new partial classes e.g. AttachmentsStorage.Sharding, CountersStorage.Sharding, etc.

ppekrol commented on src/Raven.Server/Documents/BucketStats.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace Raven.Server.Documents
+{
+    [StructLayout(LayoutKind.Explicit)]
```
Probably we need to add Pack = 1 here? I see only `long` here, but still it would not hurt?

aviv86 replied:
done

ppekrol commented on src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs:
```diff
@@ -49,6 +65,185 @@ public IEnumerable<Document> GetDocumentsByBucketFrom(DocumentsOperationContext
+        bucketStats.NumberOfDocuments += numOfDocsChanged;
+        bucketStats.LastModifiedTicks = nowTicks;
+
+        inMemoryBucketStats[bucket] = bucketStats;
```
can this be executed concurrently? probably only under write tx?

aviv86 replied:
yes, only under write tx

ppekrol commented on src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs:
```diff
@@ -49,6 +65,185 @@ public IEnumerable<Document> GetDocumentsByBucketFrom(DocumentsOperationContext
+    private static void UpdateBucketStatsInternal(Transaction tx, Slice key, long oldSize, long newSize, int numOfDocsChanged)
+    {
+        if (tx.Owner is not ShardedDocumentDatabase documentDatabase)
+            return;
```
is there a chance we can have a regular DocumentDatabase here in a normal flow? Maybe we should throw instead?

aviv86 replied:
yes, in current impl. we can have a regular DocumentDatabase here: https://github.com/ravendb/ravendb/blob/fec27c41265f53f714f2851f5bd97f4796d0812d/src/Raven.Server/Documents/DocumentsTransaction.cs#L28-#L36

ppekrol commented on src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs:
```diff
@@ -221,6 +226,11 @@ public class DynamicKeyIndexDef : AbstractTreeIndexDef
                 }
             }
 
+            public void OnIndexEntryChanged(Transaction tx, Slice key, long oldSize, long newSize)
```
maybe we should incline it aggressively?

aviv86 replied:
done

arekpalinski commented on src/Raven.Server/Documents/AttachmentsStorage.Sharding.cs:
```diff
@@ -0,0 +1,71 @@
+
+        private void UpdateBucketStatsOnPutOrDeleteStream(DocumentsOperationContext context, Slice attachmentKey, long sizeChange)
+        {
+            if (_documentDatabase is not ShardedDocumentDatabase)
```
That is a bit tricky I think. Can we throw exception here?

aviv86 replied:
we reach this code on every put/delete stream, from non-sharded document databases as well..
I'll try to change that

arekpalinski commented on src/Raven.Server/Documents/AttachmentsStorage.cs:
```diff
@@ -487,17 +468,20 @@ public void PutAttachmentStream(DocumentsOperationContext context, Slice key, Sl
             {
                 var tree = context.Transaction.InnerTransaction.CreateTree(AttachmentsSlice);
-                tree.DeleteStream(hash);
+                var size = tree.DeleteStream(hash).Size;
```
For readability let's change it to something like:

```
var deleteResult = tree.DeleteStream(hash);
var size = deleteResult .Size;
```

aviv86 replied:
done

aviv86 committed:
RavenDB-17760 - initial work on bucket stats :
- add new 'BucketStats' tree which will hold bucket info : size + num of items + last write time
- extend 'DynamicKeyIndexDef' to allow to do additional work after insert/delete/update of indexed item
- use 'AllDocsBucketAndEtag' index to update bucket info

aviv86 committed:
RavenDB-17760 : set 'OnEntryChanged' to 'DocumentsStorage.UpdateBucketStats' in all other DynamicKeyIndexDef indexes + cleanup

aviv86 committed:
RavenDB-17760 : address some PR comments
